optional/abstract_optional.py

```diff
@@ -1,47 +1,60 @@
 from abc import abstractmethod
-
 from .compatible_abc import CompatibleABC
+from typing import TypeVar, Generic, Callable, Union, NoReturn
+
+
+T = TypeVar('T')
 
 
-class AbstractOptional(CompatibleABC):
+class AbstractOptional(CompatibleABC, Generic[T]):
 
     @abstractmethod
     def is_empty(self):
+        # type: () -> bool
         pass
 
     @abstractmethod
     def get(self):
+        # type: () -> Union[T, NoReturn]
         pass
 
     @abstractmethod
     def get_or_default(self, default_value):
+        # type: (T) -> T
         pass
 
     @abstractmethod
     def get_or_raise(self, exception):
+        # type: (type) -> Union[T, NoReturn]
         pass
 
     @abstractmethod
     def if_present(self, consumer):
+        # type: (Callable) -> 'AbstractOptional'
```

Comment:
For older versions of python (pre 3.7) custom types have to be listed as a string.